Feature: flex-context per commodity as list (#2235)
FlexMeasures/flexmeasures: documentation/tut/multi-commodity.rst

                "soc-at-start": 20.0,
                "soc-min": 0.0,
                "soc-max": 100.0,
                "soc-targets": [
                    {"datetime": "2024-01-01T23:00:00+01:00", "value": 80.0}
                ],
                "power-capacity": "20 kW",
                "charging-efficiency": 0.95,
                "discharging-efficiency": 0.95
            },
            {
                "sensor": 2,
                "commodity": "gas",
                "power-capacity": "30 kW",
                "consumption-capacity": "30 kW",
                "production-capacity": "0 kW",
                "soc-usage": ["1 kW"],
                "soc-min": 0.0,
                "soc-max": 0.0,
                "soc-at-start": 0.0
            }
        ]
    }

Here, sensor ``1`` is the battery's power sensor, sensor ``2`` is the boiler's power sensor, and sensor ``3`` is the battery's instantaneous ``state-of-charge`` sensor (referenced from the battery entry so the scheduler records its charge level).

A few things to note:

- **The battery is a normal storage device** (``soc-at-start``, ``soc-min``, ``soc-max``, ``soc-targets``), tagged with ``"commodity": "electricity"``.

> - **The boiler is modelled as a fixed load.** With ``soc-min`` and ``soc-max`` both 0, it can store nothing; ``soc-usage`` of ``1 kW`` forces it to consume exactly 1 kW of gas every hour, which the optimiser cannot change. ``production-capacity`` of 0 kW means it can never feed back.

Author: ```suggestion
- **The boiler is modelled as a fixed load.** With ``soc-min`` and ``soc-max`` both 0, it can store nothing; ``soc-usage`` of ``1 kW`` forces it to consume exactly 1 kW of gas every hour, which the optimiser cannot change. ``production-capacity`` of 0 kW means it can never produce gas.
```

The prices live in the ``flex-context``. For a single commodity you would pass ``consumption-price`` and ``production-price`` directly. For **multiple commodities**, you instead provide a ``commodities`` list, one entry per commodity:

.. code-block:: json

    {
        "flex-context": {
            "commodities": [

>                 {
>                     "commodity": "electricity",
>                     "consumption-price": "100 EUR/MWh",
>                     "production-price": "100 EUR/MWh"
>                 },
>                 {
>                     "commodity": "gas",
>                     "consumption-price": "50 EUR/MWh",
>                     "production-price": "50 EUR/MWh"
>                 }
>             ]
>         }
>     }

Author: The nesting can now be decreased a little. And it may help simplify the example to leave out the gas production price.

```suggestion
    {
        "flex-context": [
            {
                "commodity": "electricity",
                "consumption-price": "100 EUR/MWh",
                "production-price": "100 EUR/MWh"
            },
            {
                "commodity": "gas",
                "consumption-price": "50 EUR/MWh"
            }
        ]
    }
```

                        "charging-efficiency": 0.95,
                        "discharging-efficiency": 0.95
                    },
                    {
                        "sensor": 2,
                        "commodity": "gas",
                        "power-capacity": "30 kW",
                        "consumption-capacity": "30 kW",
                        "production-capacity": "0 kW",
                        "soc-usage": ["1 kW"],
                        "soc-min": 0.0,
                        "soc-max": 0.0,
                        "soc-at-start": 0.0
                    }
                ],
                "flex-context": {
                    "commodities": [
                        {
                            "commodity": "electricity",
                            "consumption-price": "100 EUR/MWh",
                            "production-price": "100 EUR/MWh"
                        },
                        {
                            "commodity": "gas",
                            "consumption-price": "50 EUR/MWh",
                            "production-price": "50 EUR/MWh"
                        }
                    ]

>                 }

Author: ```suggestion
                "flex-context": [
                    {
                        "commodity": "electricity",
                        "consumption-price": "100 EUR/MWh",
                        "production-price": "100 EUR/MWh"
                    },
                    {
                        "commodity": "gas",
                        "consumption-price": "50 EUR/MWh"
                    }
                ]
```

                        "charging-efficiency": 0.95,
                        "discharging-efficiency": 0.95,
                    },
                    {
                        "sensor": 2,  # boiler power sensor
                        "commodity": "gas",
                        "power-capacity": "30 kW",
                        "consumption-capacity": "30 kW",
                        "production-capacity": "0 kW",
                        "soc-usage": ["1 kW"],
                        "soc-min": 0.0,
                        "soc-max": 0.0,
                        "soc-at-start": 0.0,
                    },
                ],
                flex_context={
                    "commodities": [
                        {
                            "commodity": "electricity",
                            "consumption-price": "100 EUR/MWh",
                            "production-price": "100 EUR/MWh",
                        },
                        {
                            "commodity": "gas",
                            "consumption-price": "50 EUR/MWh",
                            "production-price": "50 EUR/MWh",
                        },
                    ]

>                 },

Author: We need to start the flexmeasures-client PR for this, actually!

```suggestion
                flex_context=[
                    {
                        "commodity": "electricity",
                        "consumption-price": "100 EUR/MWh",
                        "production-price": "100 EUR/MWh",
                    },
                    {
                        "commodity": "gas",
                        "consumption-price": "50 EUR/MWh",
                    },
                ],
```

            )

The scheduler returns one schedule per device (stored on sensors ``1`` and ``2``) and a single commitment-cost result that breaks the cost down per commodity.

What to expect
==============

The asset chart shows both commodities together, with the battery's stock level in between:

.. image:: https://github.com/FlexMeasures/screenshots/raw/main/tut/multi-commodity.png
    :align: center
    :alt: Asset-level chart of the hybrid site, showing battery power, battery state of charge, and the gas boiler.
|

Reading the chart top to bottom:

> - **Battery power (electricity)** charges at its full 20 kW for the first three hours, then makes one partial-power step to land exactly on the 80 kWh target, and sits idle for the rest of the day. In the final hour it discharges at −20 kW. Because the electricity price is flat, there is no cheaper window to wait for, so it simply charges as early as possible (``prefer-charging-sooner`` is on by default).

Author: ```suggestion
- **Battery power (electricity)** charges at its full 20 kW for the first three hours, then makes one partial-power step, which compensates for its charging efficiency losses to land exactly on the 80 kWh target, and then sits idle for the rest of the day. In the final hour it discharges at −20 kW. Because the electricity price is flat, there is no cheaper window to wait for, so it simply charges as early as possible (``prefer-charging-sooner`` is on by default).
```

- **Gas boiler (gas)** runs at exactly 1 kW every single hour. The ``soc-usage`` field makes this a fixed load that the optimiser cannot shift — its only effect on the result is the gas cost it incurs.

The schedules match the cost figures reported by the scheduler:

.. code-block:: text

    Electricity (battery)
      Net charge needed : 80 kWh − 20 kWh        = 60 kWh stored
      Grid draw         : 60 kWh ÷ 0.95          = 63.16 kWh
      Charge cost       : 63.16 kWh × 100 EUR/MWh ≈  6.32 EUR
      Discharge credit  : 20 kWh × 100 EUR/MWh   = −2.00 EUR
      Net electricity                            ≈  4.32 EUR

    Gas (boiler)
      Consumption       : 1 kW × 24 h            = 24 kWh
      Gas cost          : 0.024 MWh × 50 EUR/MWh =  1.20 EUR

    Total                                        =  5.52 EUR

The commitment-cost result keeps these as separate entries — ``electricity net energy`` (≈ 4.32 EUR) and ``gas net energy`` (1.20 EUR) — so you can always see how much each commodity contributed.

> Because the gas price (50 EUR/MWh) is half the electricity price, serving the constant baseline with gas rather than electricity is the cheaper choice for that part of the load.

Author: Hmmm, there is no shared baseline load in this example system, though.

I suggest removing this sentence, unless there is an alternative point to make.

```suggestion
```
